Framework/Core/src/CCDBHelpers.cxx

```diff
@@ -12,24 +12,110 @@
+    std::string path;
+    int64_t cacheId;
+    std::string etag;
+  };
+
+  std::unordered_map<std::string, std::string> mapURL2UUID;
+  std::unordered_map<std::string, DataAllocator::CacheId> mapURL2DPLCache;
+
+
+  o2::ccdb::CcdbApi api;
+  std::vector<OutputRoute> routes;
+  std::map<int64_t, CCDBCacheInfo> cache;
+};
+
 AlgorithmSpec CCDBHelpers::fetchFromCCDB()
 {
-  return adaptStateful([](DeviceSpec const& spec) { 
+  return adaptStateful([](ConfigParamRegistry const&options, DeviceSpec const& spec) { 
+      std::shared_ptr<CCDBFetcherHelper> helper = std::make_shared<CCDBFetcherHelper>();
+      auto backend = options.get<std::string>("condition-backend");
+      LOGP(info, "CCDB Backend at: {}", backend);
+      helper->api.init(options.get<std::string>("condition-backend"));
+
       for (auto &route : spec.outputs) {
         if (route.matcher.lifetime != Lifetime::Condition) {
           continue;
         }
+        helper->routes.push_back(route);
         LOGP(info, "The following route is a condition {}", route.matcher);
         for (auto& metadata : route.matcher.metadata) {
-          LOGP(info, "- {}", metadata.name);
+          if (metadata.type == VariantType::String) { 
+            LOGP(info, "- {}: {}", metadata.name, metadata.defaultValue);
+          }
         }
       }
       
-      return adaptStateless([](DataAllocator& ctx) {
+      return adaptStateless([helper](RawDeviceService &device, DataTakingContext& dtc, DataAllocator& allocator, TimingInfo &timingInfo) {
+          LOGP(info, "Fetching objects. Run: {}. OrbitResetTime: {}", dtc.runNumber, dtc.orbitResetTime);
+
+          for (auto &route : helper->routes) {
+            LOGP(info, "Fetching object for route {}", route.matcher);
+
+            auto concrete = DataSpecUtils::asConcreteDataMatcher(route.matcher);
+            Output output{concrete.origin, concrete.description, concrete.subSpec, route.matcher.lifetime};
+            auto &&v = allocator.makeVector<char>(output);
+            std::map<std::string, std::string> metadata;
+            std::map<std::string, std::string> headers;
+            std::string path = "";
+            std::string etag = "";
+            std::string createdNotBefore = "0";
+            std::string createdNotAfter = "0";
+            for (auto& meta : route.matcher.metadata) {
+              if (meta.name == "ccdb-path") {
+                path = meta.defaultValue.get<std::string>();
+                break;
+              }
+            }
+            const auto url2uuid = helper->mapURL2UUID.find(path);
+            if (url2uuid != helper->mapURL2UUID.end()) {
+              etag = url2uuid->second;
+            }
+
+            LOGP(info, "{} {} {}\n", timingInfo.timeslice, timingInfo.tfCounter, timingInfo.creation);
+            helper->api.loadFileToMemory(v, path, metadata, timingInfo.timeslice, &headers, etag, createdNotAfter, createdNotBefore);
+            if ((headers.count("Error") != 0) || !v.size()) {
```

Comment:
I think `v.size()==0` should be treated as an error only if the ETAG was empty (cache was not filled yet), with ETAG for the valid object the `v` must be empty.

Reply:
done.